physical/etcd.go

```diff
@@ -167,3 +185,247 @@ func (b *EtcdBackend) nodePath(key string) string {
 func (b *EtcdBackend) nodePathDir(key string) string {
 	return filepath.Join(b.path, key) + "/"
 }
+
+// nodePathLock returns an etcd directory path used specifically for semaphore
+// indicies based on the given key.
+func (b *EtcdBackend) nodePathLock(key string) string {
+	return filepath.Join(b.path, filepath.Dir(key), EtcdNodeLockPrefix+filepath.Base(key)+"/")
+}
+
+// Lock is used for mutual exclusion based on the given key.
+func (c *EtcdBackend) LockWith(key, value string) (Lock, error) {
+	return &EtcdLock{
+		client:          c.client,
+		value:           value,
+		semaphoreDirKey: c.nodePathLock(key),
+	}, nil
+}
+
+// EtcdLock emplements a lock using and etcd backend.
+type EtcdLock struct {
+	client                               *etcd.Client
+	value, semaphoreDirKey, semaphoreKey string
+	lock                                 sync.Mutex
+}
+
+// addSemaphoreKey aquires a new ordered semaphore key.
+func (c *EtcdLock) addSemaphoreKey() (string, uint64, error) {
+
+	// CreateInOrder is an atomic operation that can be used to enqueue a
+	// request onto a semaphore. In the rest of the comments, we refer to the
+	// resulting key as a "semaphore key".
+	// https://coreos.com/etcd/docs/2.0.8/api.html#atomically-creating-in-order-keys
+	response, err := c.client.CreateInOrder(c.semaphoreDirKey, c.value, EtcdLockTTL)
+	if err != nil {
+		return "", 0, err
+	}
+
+	return response.Node.Key, response.EtcdIndex, nil
+}
+
+// getSemaphoreKey determines which semaphore key holder has aquired the lock
+// and its value.
+func (c *EtcdLock) getSemaphoreKey() (string, string, uint64, error) {
+
+	// Get the list of waiters in order to see if we are next.
+	response, err := c.client.Get(c.semaphoreDirKey, true, false)
+	if err != nil {
+		return "", "", 0, err
+	}
+
+	// Make sure the list isn't empty.
+	if response.Node.Nodes.Len() == 0 {
+		return "", "", response.EtcdIndex, nil
+	}
+
+	return response.Node.Nodes[0].Key, response.Node.Nodes[0].Value, response.EtcdIndex, nil
+}
+
+// isHeld determines if we are the current holders of the lock.
+func (c *EtcdLock) isHeld() (bool, error) {
+
+	if c.semaphoreKey == "" {
+		return false, nil
+	}
+
+	currentSemaphoreKey, _, _, err := c.getSemaphoreKey()
+	if err != nil {
+		return false, err
+	}
+
+	return c.semaphoreKey == currentSemaphoreKey, nil
+}
+
+func (c *EtcdLock) assertHeld() error {
+
+	held, err := c.isHeld()
+	if err != nil {
+		return err
+	}
+
+	if !held {
+		return EtcdLockNotHeldError
+	}
+
+	return nil
+}
+
+func (c *EtcdLock) assertNotHeld() error {
+
```

Comment:
For style, can you remove these blank lines after the start of the method and before the return?